### PR TITLE
Use reference counting for DnsQueryCandidate

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: src/resolve/resolved-dns-query.c [hunk 1]
```diff
@@ -38,19 +38,16 @@ static int dns_query_candidate_new(DnsQueryCandidate **ret, DnsQuery *q, DnsScop
 }
 
 static void dns_query_candidate_stop(DnsQueryCandidate *c) {
-        DnsTransaction *t;
-
         assert(c);
 
-        while ((t = set_steal_first(c->transactions))) {
+        for (DnsTransaction *t; (t = set_steal_first(c->transactions)); ) {
```
Comment (poettering) [inline src/resolve/resolved-dns-query.c]: that doesn't look like an improvement. typically the second expression in the for () is side-effect free. It's really weird to misuse a for() loop as a while() loop here...

File: src/resolve/resolved-dns-query.c [hunk 2]
```diff
@@ -127,36 +130,30 @@ static int dns_query_candidate_add_transaction(DnsQueryCandidate *c, DnsResource
 }
 
 static int dns_query_candidate_go(DnsQueryCandidate *c) {
+        _cleanup_(dns_query_candidate_unrefp) DnsQueryCandidate *keep_c;
```
Comment (poettering) [inline src/resolve/resolved-dns-query.c]: i'd really initialize this to zero here, even if we override it right after. having non-initialized `_cleanup_` variables have the smell of being a bug...

File: src/resolve/resolved-dns-query.c (hunk 1 above)
Comment (keszybz) [reply src/resolve/resolved-dns-query.c]: I'll drop that chunk.

File: src/resolve/resolved-dns-query.c (hunk 2 above)
Comment (keszybz) [reply src/resolve/resolved-dns-query.c]: Ack.

**Comment by keszybz** [comment]
Updated as suggested.

**Comment by DaanDeMeyer** [comment]
i386 timed out again

**Comment by po5857** [comment]
I'm afraid this fix was incomplete.  I am seeing random corruption of glibc memory structures in 246.13, leading to random backtraces which are largely useless.  So I ran systemd-resolved in valgrind and got the below info.

My assessment could be flawed, but I notice that in dns_query_cname_redirect we aren't actually taking a ref on the candidates, but are later freeing them in that function.   Are we missing a dns_query_candidate_ref there somewhere in the chain?  

Let me know if I can do something to assist here, as on one particular vm, this is reproducible within seconds of starting systemd-resolved with caching enabled (for unclear reasons).  

==1055939== Invalid read of size 8
==1055939==    at 0x13AF97: dns_query_candidate_go (resolved-dns-query.c:145)
==1055939==    by 0x13B403: dns_query_candidate_notify (resolved-dns-query.c:277)
==1055939==    by 0x141908: dns_transaction_complete (resolved-dns-transaction.c:351)
==1055939==    by 0x1435B2: dns_transaction_process_dnssec (resolved-dns-transaction.c:839)
==1055939==    by 0x1441C7: dns_transaction_process_reply (resolved-dns-transaction.c:1172)
==1055939==    by 0x1445C3: on_dns_packet (resolved-dns-transaction.c:1224)
==1055939==    by 0x4ACC1B8: source_dispatch (sd-event.c:3181)
==1055939==    by 0x4ACDBEF: sd_event_dispatch (sd-event.c:3620)
==1055939==    by 0x4ACE0C6: sd_event_run (sd-event.c:3678)
==1055939==    by 0x4ACE29D: sd_event_loop (sd-event.c:3700)
==1055939==    by 0x1168A5: run (resolved.c:92)
==1055939==    by 0x116A06: main (resolved.c:99)
==1055939==  Address 0x72c4390 is 32 bytes inside a block of size 72 free'd
==1055939==    at 0x483A9F5: free (vg_replace_malloc.c:538)
==1055939==    by 0x13A37C: mfree.lto_priv.4 (alloc-util.h:48)
==1055939==    by 0x13AAA3: dns_query_candidate_free (resolved-dns-query.c:66)
==1055939==    by 0x13AB91: dns_query_candidate_unref (resolved-dns-query.c:69)
==1055939==    by 0x13B585: dns_query_unref_candidates (resolved-dns-query.c:311)
==1055939==    by 0x13D685: dns_query_cname_redirect (resolved-dns-query.c:933)
==1055939==    by 0x13D98F: dns_query_process_cname (resolved-dns-query.c:978)
==1055939==    by 0x123B93: bus_method_resolve_hostname_complete (resolved-bus.c:154)
==1055939==    by 0x13C41D: dns_query_complete (resolved-dns-query.c:495)
==1055939==    by 0x13D1EE: dns_query_accept (resolved-dns-query.c:825)
==1055939==    by 0x13D31A: dns_query_ready (resolved-dns-query.c:858)
==1055939==    by 0x13B422: dns_query_candidate_notify (resolved-dns-query.c:287)
==1055939==  Block was alloc'd at
==1055939==    at 0x483BAE9: calloc (vg_replace_malloc.c:760)
==1055939==    by 0x13A637: dns_query_candidate_new (resolved-dns-query.c:24)
==1055939==    by 0x13C571: dns_query_add_candidate (resolved-dns-query.c:515)
==1055939==    by 0x13CD3A: dns_query_go (resolved-dns-query.c:689)
==1055939==    by 0x1247FE: bus_method_resolve_hostname (resolved-bus.c:363)
==1055939==    by 0x4A5CFA0: method_callbacks_run (bus-objects.c:415)
==1055939==    by 0x4A601F1: object_find_and_run (bus-objects.c:1325)
==1055939==    by 0x4A60870: bus_process_object (bus-objects.c:1445)
==1055939==    by 0x4A7C6BE: process_message (sd-bus.c:2777)
==1055939==    by 0x4A7C8B4: process_running.lto_priv.0 (sd-bus.c:2819)
==1055939==    by 0x4A7D4DD: bus_process_internal (sd-bus.c:3039)
==1055939==    by 0x4A7D5FC: sd_bus_process (sd-bus.c:3066)
==1055939== 
==1055939== Invalid read of size 4
==1055939==    at 0x13AB33: dns_query_candidate_unref (resolved-dns-query.c:69)
==1055939==    by 0x13A549: dns_query_candidate_unrefp (resolved-dns-query.h:108)
==1055939==    by 0x13AFDC: dns_query_candidate_go (resolved-dns-query.c:133)
==1055939==    by 0x13B403: dns_query_candidate_notify (resolved-dns-query.c:277)
==1055939==    by 0x141908: dns_transaction_complete (resolved-dns-transaction.c:351)
==1055939==    by 0x1435B2: dns_transaction_process_dnssec (resolved-dns-transaction.c:839)
==1055939==    by 0x1441C7: dns_transaction_process_reply (resolved-dns-transaction.c:1172)
==1055939==    by 0x1445C3: on_dns_packet (resolved-dns-transaction.c:1224)
==1055939==    by 0x4ACC1B8: source_dispatch (sd-event.c:3181)
==1055939==    by 0x4ACDBEF: sd_event_dispatch (sd-event.c:3620)
==1055939==    by 0x4ACE0C6: sd_event_run (sd-event.c:3678)
==1055939==    by 0x4ACE29D: sd_event_loop (sd-event.c:3700)
==1055939==  Address 0x72c4370 is 0 bytes inside a block of size 72 free'd
==1055939==    at 0x483A9F5: free (vg_replace_malloc.c:538)
==1055939==    by 0x13A37C: mfree.lto_priv.4 (alloc-util.h:48)
==1055939==    by 0x13AAA3: dns_query_candidate_free (resolved-dns-query.c:66)
==1055939==    by 0x13AB91: dns_query_candidate_unref (resolved-dns-query.c:69)
==1055939==    by 0x13B585: dns_query_unref_candidates (resolved-dns-query.c:311)
==1055939==    by 0x13D685: dns_query_cname_redirect (resolved-dns-query.c:933)
==1055939==    by 0x13D98F: dns_query_process_cname (resolved-dns-query.c:978)
==1055939==    by 0x123B93: bus_method_resolve_hostname_complete (resolved-bus.c:154)
==1055939==    by 0x13C41D: dns_query_complete (resolved-dns-query.c:495)
==1055939==    by 0x13D1EE: dns_query_accept (resolved-dns-query.c:825)
==1055939==    by 0x13D31A: dns_query_ready (resolved-dns-query.c:858)
==1055939==    by 0x13B422: dns_query_candidate_notify (resolved-dns-query.c:287)
==1055939==  Block was alloc'd at
==1055939==    at 0x483BAE9: calloc (vg_replace_malloc.c:760)
==1055939==    by 0x13A637: dns_query_candidate_new (resolved-dns-query.c:24)
==1055939==    by 0x13C571: dns_query_add_candidate (resolved-dns-query.c:515)
==1055939==    by 0x13CD3A: dns_query_go (resolved-dns-query.c:689)
==1055939==    by 0x1247FE: bus_method_resolve_hostname (resolved-bus.c:363)
==1055939==    by 0x4A5CFA0: method_callbacks_run (bus-objects.c:415)
==1055939==    by 0x4A601F1: object_find_and_run (bus-objects.c:1325)
==1055939==    by 0x4A60870: bus_process_object (bus-objects.c:1445)
==1055939==    by 0x4A7C6BE: process_message (sd-bus.c:2777)
==1055939==    by 0x4A7C8B4: process_running.lto_priv.0 (sd-bus.c:2819)
==1055939==    by 0x4A7D4DD: bus_process_internal (sd-bus.c:3039)
==1055939==    by 0x4A7D5FC: sd_bus_process (sd-bus.c:3066)
==1055939== 
Assertion 'p->n_ref > 0' failed at src/resolve/resolved-dns-query.c:69, function dns_query_candidate_unref(). Aborting.
==1055939== 
==1055939== Process terminating with default action of signal 6 (SIGABRT): dumping core
==1055939==    at 0x50AB9D5: raise (raise.c:50)
==1055939==    by 0x50948A3: abort (abort.c:79)
==1055939==    by 0x49E7FD8: log_assert_failed_realm (log.c:861)
==1055939==    by 0x13AB65: dns_query_candidate_unref (resolved-dns-query.c:69)
==1055939==    by 0x13A549: dns_query_candidate_unrefp (resolved-dns-query.h:108)
==1055939==    by 0x13AFDC: dns_query_candidate_go (resolved-dns-query.c:133)
==1055939==    by 0x13B403: dns_query_candidate_notify (resolved-dns-query.c:277)
==1055939==    by 0x141908: dns_transaction_complete (resolved-dns-transaction.c:351)
==1055939==    by 0x1435B2: dns_transaction_process_dnssec (resolved-dns-transaction.c:839)
==1055939==    by 0x1441C7: dns_transaction_process_reply (resolved-dns-transaction.c:1172)
==1055939==    by 0x1445C3: on_dns_packet (resolved-dns-transaction.c:1224)
==1055939==    by 0x4ACC1B8: source_dispatch (sd-event.c:3181)

**Pull request merged**